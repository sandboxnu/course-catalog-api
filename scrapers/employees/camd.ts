/*
 * This file is part of Search NEU and licensed under AGPL3.
 * See the license file in the root folder for details.
 */

import cheerio from "cheerio";

import macros from "../../utils/macros";
import linkSpider from "../linkSpider";
import Request from "../request";
import cache from "../cache";
import {
  parseNameWithSpaces,
  standardizeEmail,
  standardizePhone,
} from "./util";
import { Employee } from "../../types/types";

const request = new Request("Camd");

// Scrapes all the faculty and staff info from camd.northeastern.edu

class Camd {
  // Given a list of elements, this will return the text from all the elements that are text elements
  // Each text will be in its own index in the array.
  getShallowText(elements: cheerio.Element[]): string[] {
    const retVal: string[] = [];
    elements.forEach((element) => {
      if (element.type !== "text") {
        return;
      }

      const text = element.data.trim();
      if (text.length > 0) {
        retVal.push(text);
      }
    });

    return retVal;
  }

  parseDetailPage(url: string, body: string | Buffer): Employee {
    const obj: Partial<Employee> = { url: url };

    const $ = cheerio.load(body);

    // Name of person
    obj.name = $(
      "#main > div.pagecenter > div > div > div > div > div.col10.last.right > h1.entry-title"
    )
      .text()
      .trim()
      .split(",")[0];

    // Parse the first name and the last name from the given name
    const { firstName, lastName } = parseNameWithSpaces(obj.name);

    if (firstName && lastName) {
      obj.firstName = firstName;
      obj.lastName = lastName;
    }

    obj.image = $(
      "#main > div.pagecenter > div > div > div > div > div.col5 > img.wp-post-image"
    ).attr("src");
    if (obj.image) {
      obj.image = obj.image.trim();
    }

    // Primary Role
    // "Associate Professor – Design, Interactive Media"
    let primaryRole = $(
      "#main > div.pagecenter > div > div > div > div > div.col10 > p.introp"
    )
      .text()
      .trim()
      .split(";")[0];
    if (primaryRole.length > 35) {
      // Two different types of dash character
      primaryRole = primaryRole.split(" - ")[0].split(" – ")[0];
    }
    obj.primaryRole = primaryRole;

    // Phone number and office location are just both in a <p> element separated by <br>.
    // Dump all the text and then figure out where the phone and office is.
    const mainElement = $(
      "#main div.pagecenter div.gdcenter div.col16 > div.col5 > p.smallp"
    )[0] as cheerio.TagElement;
    const descriptionElements = mainElement.children;

    let email = $(
      "#main > div.pagecenter > div > div > div > div:nth-child(1) > div.col5 > p > a"
    )
      .text()
      .trim();
    email = standardizeEmail(email);
    if (email) {
      obj.emails = [email];
    }

    const texts = this.getShallowText(descriptionElements);

    texts.forEach((text) => {
      text = text.trim();
      const possiblePhone = standardizePhone(text);
      if (possiblePhone) {
        if (obj.phone) {
          macros.log("Duplicate phone?", obj.phone, possiblePhone);
        }

        obj.phone = possiblePhone;
      } else if (text.match(/[\w\d-.]+@[\w\d-.]+/) && !obj.emails) {
        // If the email was not hyperlinked, it would not be picked up by the prior email parsing and instead would appear here.
        macros.warn("Parsing plain text as email:", text);
        obj.emails = [text];
      } else if (text.length > 3) {
        // If phone did not match, check office.
        if (text.startsWith("Office: ")) {
          text = text.slice("Office: ".length);
        }

        if (obj.officeRoom) {
          macros.log(
            "Two matches for office, keeping the longer one",
            obj.office,
            text
          );

          // Only update the office if the new office is longer.
          // This rarely happens, but the longer the string is the more likely it is to be an office location.
          // In all of CAMD, there are only 2 instance where this helps
          if (obj.officeRoom.length < text.length) {
            obj.officeRoom = text;
          }

          return;
        }

        obj.officeRoom = text;
      } else {
        macros.log("Warn: unknown prop in description", text);
      }
    });

    return obj as Employee;
  }

  async main(): Promise<Employee[]> {
    if (macros.DEV && require.main !== module) {
      const devData = await cache.get(
        macros.DEV_DATA_DIR,
        this.constructor.name,
        "main"
      );
      if (devData) {
        return devData as Employee[];
      }
    }

    const startingLinks = [
      "https://camd.northeastern.edu/architecture/faculty-staff",
      "https://camd.northeastern.edu/artdesign/faculty-staff",
      "https://camd.northeastern.edu/commstudies/faculty-staff",
      "https://camd.northeastern.edu/gamedesign/faculty-staff",
      "https://camd.northeastern.edu/journalism/faculty-staff",
      "https://camd.northeastern.edu/mscr/faculty-staff",
      "https://camd.northeastern.edu/music/faculty-staff",
      "https://camd.northeastern.edu/theatre/faculty-staff",
    ];

    const urls = await linkSpider.main(startingLinks);

    const profileUrls = [];

    // Filter all the urls found to just profile urls
    // 'https://camd.northeastern.edu/artdesign/people/magy-seif-el-nasr-2/',
    urls.forEach((url) => {
      if (
        url.match(
          /https:\/\/camd.northeastern.edu\/(architecture|artdesign|commstudies|gamedesign|journalism|mscr|music|theatre)\/people\/[\w\d-]+\/?/i
        )
      ) {
        profileUrls.push(url);
      }
    });

    const promises: Promise<Employee>[] = [];

    profileUrls.forEach((url) => {
      promises.push(
        request.get(url).then((response) => {
          return this.parseDetailPage(url, response.body);
        })
      );
    });

    const people = await Promise.all(promises);

    if (macros.DEV) {
      await cache.set(
        macros.DEV_DATA_DIR,
        this.constructor.name,
        "main",
        people
      );
      macros.log(people.length, "camd people saved to the cache file!");
    }

    return people;
  }
}

const instance = new Camd();

if (require.main === module) {
  instance.main();
}

export default instance;
