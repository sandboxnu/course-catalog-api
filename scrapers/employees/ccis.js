/*
 * This file is part of Search NEU and licensed under AGPL3.
 * See the license file in the root folder for details.
 */

import cheerio from "cheerio";

import Request from "../request";
import macros from "../../utils/macros";
import cache from "../cache";
import {
  standardizeEmail,
  standardizePhone,
  parseGoogleScolarLink,
  parseNameWithSpaces,
} from "./util";

const request = new Request("CCIS");

// Scrapes professor info from
// http://www.ccis.northeastern.edu/people-view-all/

// Scraped info:
// "name": "Amal Ahmed",
// "url": "http://www.ccis.northeastern.edu/people/amal-ahmed/",
// "positions": ["Assistant Professor"],
// "email": "amal@ccs.neu.edu",
// "phone": "6173732076",
// "office": "440 Huntington Avenue\r\n328 West Village H\r\nBoston, MA 02115",

// This varies a lot. Some have links to their page on other schools, some have self hosted sites, etc
// "personalSite": "http://www.ccs.neu.edu/home/amal/",

// Link to some google site. Get url with:
// 'https://scholar.google.com/citations?user=' + googleScholarId + '&hl=en&oi=ao'
// "googleScholarId": "Y1C007wAAAAJ",
// "bigPictureUrl": "http://www.ccis.northeastern.edu/wp-content/uploads/2016/03/Amal-Ahmed-hero-image.jpg"

// There were a few people that put non-google urls on the google url field.
// These links are ignored at the moment.
// They could be used in place of the personal website url, but there were only a few professors and most of them had personal website links too.

// TODO: also could clean up the first name and last name parsing.
// For 3 people in all of CCIS, they have names like [Panagiotos (Pete) Manolios]
// Currently the first word is used as the first name and the last one is used as the last name
// But we could use the middle one as the first name if it starts and ends with parens

class NeuCCISFaculty {
  parsePeopleList(resp) {
    const $ = cheerio.load(resp.body);

    const output = [];

    const peopleElements = $(
      "div.letter > div.people > article.people-directory"
    );

    for (let i = 0; i < peopleElements.length; i++) {
      const $personElement = $(peopleElements[i]);
      const obj = {};

      obj.name = $("h3.person-name", $personElement).text().trim();

      // Parse the first name and the last name from the given name
      const { firstName, lastName } = parseNameWithSpaces(obj.name);

      if (firstName && lastName) {
        obj.firstName = firstName;
        obj.lastName = lastName;
      }

      // Link to profile.
      obj.url = $("h3.person-name > a", $personElement).attr("href").trim();

      // positions at neu (eg PhD Student)
      const positionElements = $(
        "div.position-list > span.position",
        $personElement
      );
      if (positionElements.length > 0) {
        const positions = [];
        for (let j = 0; j < positionElements.length; j++) {
          positions.push($(positionElements[j]).text().trim());
        }

        // Hold off on keeping the positions array for now. Need to ensure it is the same schema as COE (and other parser that are scraping positions).
        obj.primaryRole = positions[0];
      }

      // email
      const emailElements = $(
        "div.contact-info > div.email > a",
        $personElement
      );

      // also email
      if (emailElements.length > 0) {
        const email = standardizeEmail(emailElements.text().trim());
        const mailto = standardizeEmail(emailElements.attr("href").trim());

        if (!mailto || !email || mailto !== email) {
          macros.log("Warning: bad emails?", email, mailto, obj.name);
        } else {
          obj.emails = [email];
        }
      }

      // Phone
      const phoneElements = $(
        "div.contact-info > div.phone > a",
        $personElement
      );
      if (phoneElements.length > 0) {
        let phone = phoneElements.text().trim();

        let tel = phoneElements.attr("href").trim();

        tel = standardizePhone(tel);
        phone = standardizePhone(phone);

        if (tel || phone) {
          if (!tel || !phone || tel !== phone) {
            macros.log("phone tel mismatch", tel, phone, obj);
          } else {
            obj.phone = phone;
          }
        }
      }

      ["name", "url", "emails"].forEach((attrName) => {
        if (!obj[attrName]) {
          macros.log("Missing", attrName, "for", obj.name);
        }
      });

      output.push(obj);
    }
    return output;
  }

  parseDetailpage(resp, obj = {}) {
    const $ = cheerio.load(resp.body);

    const office = $("div.contact-block > div.address > p").text();
    if (office) {
      const officeSplit = office.replace(/\r\n/gi, "\n").trim().split("\n");

      let officeRoom = officeSplit[1];

      if (officeRoom) {
        officeRoom = officeRoom.trim();

        // Need to remove trailing commas
        if (officeRoom.endsWith(",")) {
          officeRoom = officeRoom.slice(0, officeRoom.length - 1);
        }
        obj.officeRoom = officeRoom;
      }

      obj.officeStreetAddress = officeSplit[0].trim();
    }

    obj.personalSite = $(
      "div.contact-block > div.contact-links > p.personal-site > a"
    ).attr("href");
    if (obj.personalSite) {
      obj.personalSite = obj.personalSite.trim();
    }

    const googleScholarUrl = $(
      "div.contact-block > div.contact-links > p.google-scholar > a"
    ).attr("href");

    const userId = parseGoogleScolarLink(googleScholarUrl);
    if (userId) {
      obj.googleScholarId = userId;
    }

    obj.bigPictureUrl = $(
      "header.people-header > div.section-inner > img"
    ).attr("src");
    if (obj.bigPictureUrl) {
      obj.bigPictureUrl = obj.bigPictureUrl.trim();
    }

    return obj;
  }

  async main() {
    // If this is dev and this data is already scraped, just return the data.
    if (macros.DEV && require.main !== module) {
      const devData = await cache.get(
        macros.DEV_DATA_DIR,
        this.constructor.name,
        "main"
      );
      if (devData) {
        return devData;
      }
    }

    const resp = await request.get(
      "http://www.khoury.northeastern.edu/people-view-all/"
    );
    const peopleObjects = this.parsePeopleList(resp);

    const promises = [];
    const output = [];

    // Cool, parsed all of the info from the first page
    // Now scrape each profile
    peopleObjects.forEach((obj) => {
      promises.push(
        request.get(obj.url).then((personResponse) => {
          output.push(this.parseDetailpage(personResponse, obj));
          return;
        })
      );
    });

    await Promise.all(promises);

    if (macros.DEV) {
      await cache.set(
        macros.DEV_DATA_DIR,
        this.constructor.name,
        "main",
        output
      );
      macros.log(output.length, "people in ccis saved to a file!");
    }

    macros.log("done!");
    return output;
  }
}

const instance = new NeuCCISFaculty();
export default instance;

if (require.main === module) {
  instance.main();
}
