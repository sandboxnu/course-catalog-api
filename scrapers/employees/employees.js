/*
 * This file is part of Search NEU and licensed under AGPL3.
 * See the license file in the root folder for details.
 */

import htmlparser from "htmlparser2";
import domutils from "domutils";
import _ from "lodash";
import cookie from "cookie";
import he from "he";

import Request from "../request";
import cache from "../cache";
import macros from "../../utils/macros";
import { standardizeEmail, occurrences } from "./util";

const request = new Request("Employees");

// Scrapes from here: https://prod-web.neu.edu/wasapp/employeelookup/public/main.action

// TODO:
// Some of the phone numbers are not 10 digits. Might be able to guess area code, but it is not allways 215 just because some people have offices outside of Boston.
// Phone numbers that are not 10 digits are ignored right now.
// The detail page also has a name without the comma, but don't see a reason on scraping that instead of just splitting on comma and moving.
// Is that name ever different than the one with the comma?
// Their office is also in that detail page, but it seems like if is often different than the ones the professors have
// on their CCIS/COE/CSSH profile. Unsure if it would be worth scraping that too.

// Currently:
// Name is always scraped. This can vary slightly between different data sources.
// This name will never include accents, which the CCIS site does
// Also, multiple people have the same name so this can not be used as a primary key
// "name": "Hauck, Heather",

// Id of each employee. Always scraped.
// In Aug, 2019 this id was changed from a decimal to a base64 encoded binary string on NEU's site.
// "id": "%2B9Yas6J4nmbQ0e8J7Ju54Q%3D%3D",

// Phone number. Some people had them posted and others did not. Sometimes present, sometimes not.
// The same phone number can be listed as one person's phone number on this data source and a different person on a different data source.
// Don't have a great idea on how to handle that.
// "phone": "6173737821",

// Email. Sometimes present, sometimes not.
// "email": "h.hauck@northeastern.edu",

// Sometimes present, sometimes not. Often heavily abbreviated. (like this example)
// "primaryappointment": "Asst Dir &amp; Assoc Coop Coord",

// Always scraped.
// "primarydepartment": "DMSB Co-op"

// Should be done better here:
// employees.parseLettersResponse should be stateless, and just return some json from the html it processes
// and employees.parseLettersResponse should be sync, and not async
// and we could get rid of domutils too lmao

class Employee {
  constructor() {
    this.people = [];

    this.couldNotFindNameList = {};

    this.cookiePromise = null;
  }

  handleRequestResponce(body, callback) {
    const handler = new htmlparser.DomHandler(callback);
    const parser = new htmlparser.Parser(handler);
    parser.write(body);
    parser.done();
  }

  //returns a {colName:[values]} where colname is the first in the column
  //regardless if its part of the header or the first row of the body
  parseTable(table) {
    if (table.name !== "table") {
      macros.error("parse table was not given a table..");
      return null;
    }

    //includes both header rows and body rows
    const rows = domutils.getElementsByTagName("tr", table);

    if (rows.length === 0) {
      return null;
    }

    const retVal = {};
    const heads = [];

    //the headers
    rows[0].children.forEach((element) => {
      if (element.type !== "tag" || ["th", "td"].indexOf(element.name) === -1) {
        return;
      }

      const text = domutils
        .getText(element)
        .trim()
        .toLowerCase()
        .replace(/\s/gi, "");
      retVal[text] = [];
      heads.push(text);
    });

    //add the other rows
    rows.slice(1).forEach((row) => {
      let index = 0;
      row.children.forEach((element) => {
        if (
          element.type !== "tag" ||
          ["th", "td"].indexOf(element.name) === -1
        ) {
          return;
        }
        if (index >= heads.length) {
          macros.log(
            "warning, table row is longer than head, ignoring content",
            index,
            heads,
            rows
          );
          return;
        }

        retVal[heads[index]].push(domutils.getText(element).trim());

        //only count valid elements, not all row.children
        index++;
      });

      //add empty strings until reached heads length
      for (; index < heads.length; index++) {
        retVal[heads[index]].push("");
      }
    });
    return {
      rowCount: rows.length - 1,
      parsedTable: retVal,
    };
  }

  async getCookiePromise() {
    if (this.cookiePromise) {
      return this.cookiePromise;
    }

    macros.verbose("neu employee getting cookie");

    this.cookiePromise = request
      .get({
        url:
          "https://prod-web.neu.edu/wasapp/employeelookup/public/main.action",
      })
      .then((resp) => {
        // Parse the cookie from the response
        const cookieString = resp.headers["set-cookie"][0];
        const cookies = cookie.parse(cookieString);
        return cookies.JSESSIONID;
      });

    return this.cookiePromise;
  }

  hitWithLetters(lastNameStart, jsessionCookie) {
    return request.get({
      url: `https://prod-web.neu.edu/wasapp/employeelookup/public/searchEmployees.action?searchBy=Last+Name&queryType=begins+with&searchText=${lastNameStart}&deptText=&addrText=&numText=&divText=&facStaff=2`,
      headers: {
        Cookie: `JSESSIONID=${jsessionCookie}`,
      },
    });
  }

  // Given a list of things, will find the first one that is longer than 1 letter (a-z)
  findName(list) {
    for (let i = 0; i < list.length; i++) {
      const noSymbols = list[i].toLowerCase().replace(/[^0-9a-zA-Z]/gi, "");

      if (
        noSymbols.length > 1 &&
        !["ii", "iii", "jr", "sr", "dr"].includes(noSymbols)
      ) {
        return list[i];
      }
    }

    // Only log each warning once, just to not spam the macros. This method is called a lot.
    const logMatchString = list.join("");
    if (this.couldNotFindNameList[logMatchString]) {
      return null;
    }
    this.couldNotFindNameList[logMatchString] = true;

    macros.log("Could not find name from list:", list);
    return null;
  }

  // This splits the employee name by the comma in the middle and the name after the comma is the first name and the name before is the last name
  // The util function for this does not split the name by the comma, it assumes that the name is just separated by spaces.
  getFirstLastName(name) {
    if (name.match(/jr.?,/gi)) {
      name = name.replace(/, jr.?,/gi, ",");
    }

    name = _.trim(name, ",");

    if (occurrences(name, ",") !== 1) {
      macros.log("Name has != 1 commas", name);
      return null;
    }

    const splitOnComma = name.split(",");

    const beforeCommaSplit = splitOnComma[1].trim().split(" ");
    const firstName = this.findName(beforeCommaSplit);

    const afterCommaSplit = splitOnComma[0].trim().split(" ").reverse();
    const lastName = this.findName(afterCommaSplit);

    const retVal = {};
    retVal.firstName = firstName;
    retVal.lastName = lastName;
    return retVal;
  }

  parseLettersResponse(response, lastNameStart) {
    return new Promise((resolve, reject) => {
      this.handleRequestResponce(response.body, (err, dom) => {
        const elements = domutils.getElementsByTagName("table", dom);

        for (let i = 0; i < elements.length; i++) {
          const element = elements[i];

          const goal = {
            width: "100%",
          };

          if (_.isEqual(element.attribs, goal)) {
            // Delete one of the elements that is before the header that would mess stuff up
            domutils.removeElement(element.children[1].children[1]);

            const tableData = this.parseTable(element);

            if (!tableData) {
              return resolve();
            }

            const parsedTable = tableData.parsedTable;
            const rowCount = tableData.rowCount;

            macros.log("Found", rowCount, " people on page ", lastNameStart);

            for (let j = 0; j < rowCount; j++) {
              const person = {};
              const nameWithComma = he
                .decode(parsedTable.name[j])
                .split("\n\n")[0];

              if (nameWithComma.includes("Do Not Use ")) {
                macros.log("Skipping entry that says Do Not Use.");
                continue;
              }

              // Put the first name before the last name.
              // Another way to do this would be to hit the detail section of each employee and scrape the name from the title.
              const commaNameSplit = nameWithComma.split(",");

              for (let k = 0; k < commaNameSplit.length; k++) {
                commaNameSplit[k] = commaNameSplit[k].trim();
              }

              // Remove Jr and Jr.
              _.pull(commaNameSplit, "Jr", "Jr.");

              if (commaNameSplit.length > 2) {
                macros.warn(
                  "Has more than one comma skipping.",
                  commaNameSplit
                );
                person.name = nameWithComma;
              } else {
                person.name = `${commaNameSplit[1].trim()} ${commaNameSplit[0].trim()}`;
              }

              // Generate first name and last name from the name on the person
              const firstLastName = this.getFirstLastName(nameWithComma);
              if (!firstLastName) {
                continue;
              }
              const { firstName, lastName } = firstLastName;
              if (firstName && lastName) {
                person.firstName = firstName;
                person.lastName = lastName;
              }

              // The "#anchor_68409676" tags on the page are randomly generated each time the page is loaded,
              // so use this hrefparameter id instead, which should be more stable.
              // This hrefparameter param is a url encoded, base64 string.
              // We could decode it, but there's just no point.
              const idMatch = parsedTable.name[j].match(
                /.hrefparameter\s+=\s+"id=([\d\w%+/]+)";/i
              );
              if (!idMatch) {
                macros.warn(
                  "Unable to parse id, using random number",
                  nameWithComma
                );
                person.id = String(Math.random());
              } else {
                let id = idMatch[1];

                // Trim the id if it is too long. This is just a sanity check.
                // 35 is just a arbitrary decided number.
                if (id.length > 35) {
                  macros.warn("person id over 35 chars?", id);
                  id = id.slice(0, 35);
                }

                person.id = id;
              }

              let phone = parsedTable.phone[j];
              phone = phone.replace(/\D/g, "");

              // Maybe add support for guesing area code if it is ommitted and most of the other ones have the same area code
              if (phone.length === 10) {
                person.phone = phone;
              }

              // Scrape the email from the table
              const email = standardizeEmail(parsedTable.email[j]);
              if (email) {
                person.emails = [email];
              }

              // Scrape the primaryappointment
              const primaryappointment = parsedTable.primaryappointment[j];
              if (
                primaryappointment &&
                primaryappointment !== "Not Available"
              ) {
                person.primaryRole = he.decode(
                  parsedTable.primaryappointment[j]
                );
              }

              // Scrape the primary department
              person.primaryDepartment = he.decode(
                parsedTable.primarydepartment[j]
              );

              // Add it to the people list
              this.people.push(person);
            }
            return resolve();
          }
        }

        macros.error("YOOOOO it didnt find the table", response.body.length);
        macros.error(response.body);
        return reject();
      });
    });
  }

  async get(lastNameStart) {
    const jsessionCookie = await this.getCookiePromise();

    macros.verbose("neu employee got cookie", jsessionCookie);

    const response = await this.hitWithLetters(lastNameStart, jsessionCookie);

    return this.parseLettersResponse(response, lastNameStart);
  }

  async main() {
    // if this is dev and this data is already scraped, just return the data
    if (macros.DEV && require.main !== module) {
      const devData = await cache.get(
        macros.DEV_DATA_DIR,
        this.constructor.name,
        "main"
      );
      if (devData) {
        return devData;
      }
    }

    const promises = [];

    const alphabetArray = macros.ALPHABET.split("");

    for (const firstLetter of alphabetArray) {
      for (const secondLetter of alphabetArray) {
        promises.push(this.get(firstLetter + secondLetter));
      }
    }

    await Promise.all(promises);

    macros.verbose("Done all employee requests");

    if (macros.DEV) {
      await cache.set(
        macros.DEV_DATA_DIR,
        this.constructor.name,
        "main",
        this.people
      );
      macros.log(this.people.length, "employees saved to a file!");
    }

    return this.people;
  }
}

const instance = new Employee();

export default instance;

if (require.main === module) {
  instance.main();
}
