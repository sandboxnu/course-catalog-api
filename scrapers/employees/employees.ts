/*
 * This file is part of Search NEU and licensed under AGPL3.
 * See the license file in the root folder for details.
 */

import Request from "../request";
import cache from "../cache";
import macros from "../../utils/macros";
import {
  Employee,
  EmployeeRequestResponse,
  EmployeeWithId,
} from "../../types/types";
import { v4 as uuidv4 } from "uuid";

const request = new Request("Employees");

class NeuEmployee {
  people: EmployeeWithId[];

  constructor() {
    this.people = [];
  }

  parseParameter(parameter: string): string {
    // The NEU API returns 'Not Avaiable' for some fields instead of making them null
    return parameter.toLowerCase() !== "not available" ? parameter : null;
  }

  parseApiResponse(response: EmployeeRequestResponse[]): EmployeeWithId[] {
    return (
      response
        .map((employee) => {
          const email = this.parseParameter(employee.Email);

          return {
            id: email ? email : uuidv4(),
            name: `${employee.FirstName} ${employee.LastName}`,
            firstName: employee.FirstName,
            lastName: employee.LastName,
            primaryDepartment: employee.Department,
            primaryRole: this.parseParameter(employee.PositionTitle),
            phone: this.parseParameter(employee.PhoneNumber),
            emails: email ? [email] : [],
            email: email,
            officeRoom: this.parseParameter(employee.CampusAddress),
          };
        })
        // Northeastern likes testing in prod (don't we all)
        // As of 2022-03, they have 4 employees whose names are variations of "do not use"
        //    Do Not Use, Ed
        //    Do Not Use Hrm See 000499418, Do Not Use
        //    DO NOT USE, DO NOT USE
        .filter(
          (employee) => !employee.name.toLowerCase().includes("do not use")
        )
    );
  }

  /**
   * Queries the Northeastern employees API.
   */
  async queryEmployeesApi(): Promise<void> {
    // These values are taken from the API exposed here: https://nu.outsystemsenterprise.com/FSD/
    // The purpose of these values is unknown, as the API is undocumented. If requests start failing, please
    // inspect the API response at the above link & update accordingly

    // Scrapes the source code to get the X-CSRF token.
    const csrfToken = await request
      .get({
        url: "https://nu.outsystemsenterprise.com/FSD/scripts/OutSystems.js",
      })
      .then((resp) => {
        return resp.body.match(/"X-CSRFToken".*?="(.*?)"/i)[1];
      });

    const moduleVersion = await request
      .get({
        url: "https://nu.outsystemsenterprise.com/FSD/moduleservices/moduleversioninfo",
      })
      .then((resp) => resp.body["versionToken"]);

    const employeeQuery = {
      versionInfo: {
        moduleVersion,
        apiVersion: "ad_141F5PYcK3c+D5K8O+g",
      },
      viewName: "MainFlow.FacultyAndStaffDirectory",
      // Their frontend requires non-null queries, but the backend doesn't :)
      // If this changes at any point, we used to do a cartesian product of the alphabet
      // ('aa', 'ab', etc.). See https://github.com/sandboxnu/course-catalog-api/pull/97/files for more
      inputParameters: {
        Input_Name1: "",
        Input_Name2: "",
      },
    };

    const response: EmployeeRequestResponse[] = await request
      .post({
        url: "https://nu.outsystemsenterprise.com/FSD/screenservices/FSD/MainFlow/Name/ActionGetContactsByName_NonSpecificType",
        body: employeeQuery,
        json: true,
        headers: {
          "X-CSRFToken": csrfToken,
        },
      })
      .then((r) => r.body.data.EmployeeDirectoryContact.List);

    this.people = this.parseApiResponse(response);
  }

  async main(): Promise<Employee[]> {
    // if this is dev and this data is already scraped, just return the data
    if (macros.DEV && require.main !== module) {
      const devData = await cache.get(
        macros.DEV_DATA_DIR,
        this.constructor.name,
        "main"
      );
      if (devData) {
        return (devData as Employee[]).map((employee) => {
          const id = employee?.email ?? uuidv4();
          return { ...employee, id };
        });
      }
    }

    await this.queryEmployeesApi();

    macros.verbose("Done all employee requests");

    if (macros.DEV) {
      await cache.set(
        macros.DEV_DATA_DIR,
        this.constructor.name,
        "main",
        this.people
      );
      macros.log(this.people.length, "NEU employees saved.");
    }

    return this.people;
  }
}

const instance = new NeuEmployee();

export default instance;

if (require.main === module) {
  instance.main();
}
