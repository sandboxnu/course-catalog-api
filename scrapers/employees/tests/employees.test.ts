/*
 * This file is part of Search NEU and licensed under AGPL3.
 * See the license file in the root folder for details.
 */

import employeeData from "./data/employees/employee_results_BE.json";

import employees from "../employees";
import { validate } from "uuid";

// This should not be top-level, but we needed a workaround.
// https://github.com/facebook/jest/issues/11543
jest.unmock("request-promise-native");
jest.setTimeout(20_000);

describe("static data", () => {
  // Test to make sure parsing of an employees result page stays the same
  it("should be able to parse API responses", async () => {
    const employeeList = employees.parseApiResponse(employeeData);

    for (const employee of employeeList) {
      // https://stackoverflow.com/questions/46155/whats-the-best-way-to-validate-an-email-address-in-javascript
      const validId =
        employee.id.match(/\S+@\S+\.\S+/) || validate(employee.id);
      expect(validId).toBeTruthy();
      expect(employee.name.toLowerCase().includes("do not use")).toBeFalsy();

      // The employees API returns 'Not Avaiable' for some fields instead of making them null
      // This makes sure we filter them all out
      const matchesNotAvailable = (obj: any): boolean =>
        /Not Available/.test(JSON.stringify(obj));

      expect(matchesNotAvailable({ key: "Not Available" })).toBeTruthy(); // Sanity-check the regex
      expect(matchesNotAvailable(employee)).toBeFalsy();
    }

    expect(employeeList).toMatchSnapshot();
  });
});

describe("scraping employees", () => {
  it("should be able to query the API", async () => {
    await employees.queryEmployeesApi();

    // Don't use a precise number - this hits the live API, so we don't know how many there are
    expect(employees.people.length).toBeGreaterThan(1_000);

    for (const employee of employees.people) {
      // Ensure that the fields we expect to be required actually are
      expect(employee.name).toBeTruthy();
      expect(employee.firstName).toBeTruthy();
      expect(employee.lastName).toBeTruthy();
    }
  });
});
