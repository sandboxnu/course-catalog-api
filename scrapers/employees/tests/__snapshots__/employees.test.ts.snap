// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`static data should be able to parse API responses 1`] = `
Array [
  Object {
    "email": "m.abedi@northeastern.edu",
    "emails": Array [
      "m.abedi@northeastern.edu",
    ],
    "firstName": "Mehdi",
    "id": "m.abedi@northeastern.edu",
    "lastName": "Abedi",
    "name": "Mehdi Abedi",
    "officeRoom": "334 SN",
    "phone": "6173734427",
    "primaryDepartment": "Mechanical + Industrial Engineering",
    "primaryRole": "Associate Teaching Professor",
  },
  Object {
    "email": "h.abeels@northeastern.edu",
    "emails": Array [
      "h.abeels@northeastern.edu",
    ],
    "firstName": "Hannah",
    "id": "h.abeels@northeastern.edu",
    "lastName": "Abeels",
    "name": "Hannah Abeels",
    "officeRoom": "KN 120 Knowles",
    "phone": "6173735149",
    "primaryDepartment": "School Of Law-Academic",
    "primaryRole": "Asst Dir - Online & Hybrid Pro",
  },
  Object {
    "email": "s.abel@northeastern.edu",
    "emails": Array [
      "s.abel@northeastern.edu",
    ],
    "firstName": "Sherry",
    "id": "s.abel@northeastern.edu",
    "lastName": "Abel",
    "name": "Sherry Abel",
    "officeRoom": "101 N. Tyron St",
    "phone": "9802248466",
    "primaryDepartment": "School Of Nursing",
    "primaryRole": "Part-time Lecturer",
  },
  Object {
    "email": "p.abelli@northeastern.edu",
    "emails": Array [
      "p.abelli@northeastern.edu",
    ],
    "firstName": "Peter",
    "id": "p.abelli@northeastern.edu",
    "lastName": "Abelli",
    "name": "Peter Abelli",
    "officeRoom": "140 CN",
    "phone": "6173735599",
    "primaryDepartment": "Maintenance-HVAC-Dorms",
    "primaryRole": "HVAC Systems Technician",
  },
  Object {
    "email": "m.abels@northeastern.edu",
    "emails": Array [
      "m.abels@northeastern.edu",
    ],
    "firstName": "Margot",
    "id": "m.abels@northeastern.edu",
    "lastName": "Abels",
    "name": "Margot Abels",
    "officeRoom": "310 RP",
    "phone": null,
    "primaryDepartment": "Human Services",
    "primaryRole": "Assistant Teaching Professor",
  },
  Object {
    "email": "e.ackerman@northeastern.edu",
    "emails": Array [
      "e.ackerman@northeastern.edu",
    ],
    "firstName": "Elizabeth",
    "id": "e.ackerman@northeastern.edu",
    "lastName": "Ackerman",
    "name": "Elizabeth Ackerman",
    "officeRoom": "102 RB",
    "phone": null,
    "primaryDepartment": "BCHS Office of the Dean",
    "primaryRole": "Part-Time Lecturer",
  },
  Object {
    "email": "m.acosta@northeastern.edu",
    "emails": Array [
      "m.acosta@northeastern.edu",
    ],
    "firstName": "Beth",
    "id": "m.acosta@northeastern.edu",
    "lastName": "Acosta",
    "name": "Beth Acosta",
    "officeRoom": "RB 103 RB",
    "phone": "6173735177",
    "primaryDepartment": "School Of Nursing",
    "primaryRole": "Part-time Lecturer",
  },
  Object {
    "email": "m.adams@northeastern.edu",
    "emails": Array [
      "m.adams@northeastern.edu",
    ],
    "firstName": "Marybeth",
    "id": "m.adams@northeastern.edu",
    "lastName": "Adams-Khoury",
    "name": "Marybeth Adams-Khoury",
    "officeRoom": "102 RB",
    "phone": null,
    "primaryDepartment": "BCHS Office of the Dean",
    "primaryRole": "Part-Time Lecturer",
  },
  Object {
    "email": "j.adelberg@northeastern.edu",
    "emails": Array [
      "j.adelberg@northeastern.edu",
    ],
    "firstName": "Jeff",
    "id": "j.adelberg@northeastern.edu",
    "lastName": "Adelberg",
    "name": "Jeff Adelberg",
    "officeRoom": "360 Huntington Ave",
    "phone": null,
    "primaryDepartment": "Theater",
    "primaryRole": "Part-Time Lecturer",
  },
]
`;
